Feat/native meta contract

carpawell opened the pull request:
New application configs example: https://github.com/nspcc-dev/neofs-dev-env/pull/334

carpawell commented on cmd/neofs-node/meta_new.go:
```diff
@@ -0,0 +1,164 @@
+
+	p2pCfg := c.appCfg.Meta.P2P
+
+	var chainCfg = config.Config{
```
i would like to be validated on what exact minimum config is enough for an RPC node to start. see `applySidechainDefaults` also below

carpawell commented on pkg/innerring/internal/metachain/contracts/meta.go:
```diff
@@ -0,0 +1,99 @@
+	if ic.Tx != nil {
+		txHash = ic.Tx.Hash().StringLE()
+	}
+	ic.Log.Info(runtime.SystemRuntimeLogMessage, append(fields,
```
do not know if there are any agreements on logs. there is a system call about them, not sure if it is ok to log smth in a free manner

AnnaShaleva commented on cmd/neofs-node/meta_new.go:
```diff
@@ -0,0 +1,164 @@
+	}
+	if cfg.ApplicationConfiguration.P2P.AttemptConnPeers == 0 {
+		cfg.ApplicationConfiguration.P2P.AttemptConnPeers = cfg.ApplicationConfiguration.P2P.MinPeers + 10
+	}
```
As far as some other fields mentioned in `applySidechainDefaults`. So you may safely omit these checks if NeoGo's defaults are OK for you.

AnnaShaleva commented on pkg/innerring/internal/metachain/contracts/meta.go:
```diff
@@ -0,0 +1,99 @@
+	desc := native.NewDescriptor("submitObjectPut", smartcontract.VoidType,
+		manifest.NewParameter("metaInformation", smartcontract.ByteArrayType),
+		manifest.NewParameter("signatures", smartcontract.ArrayType))
+	md := native.NewMethodAndPrice(m.submitObjectPut, 0, callflag.States|callflag.AllowNotify)
```
Is it free?

AnnaShaleva commented on pkg/innerring/internal/metachain/contracts/method_objects.go:
```diff
@@ -0,0 +1,218 @@
+	}
+
+	return nil
+}
```
Use `(INEO).CheckCommittee` instead, that's it.

AnnaShaleva commented on pkg/innerring/internal/metachain/contracts/method_containers.go:
```diff
@@ -0,0 +1,59 @@
+	}
+
+	ic.DAO.DeleteStorageItem(m.ID, append([]byte{metaContainersPrefix}, cID...))
+	logMessage(ic, "removed meta container", zap.String("cID", base58.Encode(cID)))
```
Yep, either send a proper notification or don't log at all.

AnnaShaleva commented on pkg/innerring/internal/metachain/contracts/meta_test.go:
```diff
@@ -0,0 +1,262 @@
+		case *native.NEO:
+			neoContract = contract.(native.INEO)
+			newContracts = append(newContracts, neoContract)
+		case *native.Management, *native.Ledger, *native.GAS, *native.Policy, *native.Designate, *native.Notary:
```
We're supposed to have a custom (and simplified) Neo/Gas implementations. Also, do you need Notary at all for this chain?

AnnaShaleva replied:
I suppose we need a separate issue to simplify the set of native contracts for meta chain.

AnnaShaleva commented on pkg/innerring/internal/metachain/contracts/method_containers.go:
```diff
@@ -0,0 +1,59 @@
+	err := m.checkCommitteeWitness(ic)
+	if err != nil {
+		panic(err)
+	}
```
```suggestion
	m.neo.CheckCommittee(ic)
```

carpawell commented on cmd/neofs-node/meta_new.go:
```diff
@@ -0,0 +1,164 @@
+			Genesis: config.Genesis{
+				MaxTraceableBlocks:          fsChainProtocol.MaxTraceableBlocks,
+				MaxValidUntilBlockIncrement: fsChainProtocol.MaxValidUntilBlockIncrement,
+				Roles:                       nil, // not needed?
```
> Also, don't you use Notaries for this chain?

i do, but the notary service starts only on the IR side, and this is SN's RPC node that does not receive notary requests. looks like this is a mistake, but for some reason it starts and fetches blocks ok

AnnaShaleva commented on pkg/innerring/internal/metachain/contracts/method_objects.go:
```diff
@@ -0,0 +1,218 @@
+	if !ok {
+		panic(fmt.Errorf("unexpected deserialized meta information value: expected %T, %T, given", metaInfo, metaInfoSI.Value()))
+	}
+	cID, err := requiredInMap(metaInfo, "cid").TryBytes()
```
Maps are ordered, you may access `cid` via index.

AnnaShaleva replied:
> an array should be used instead if we rely on the order

Agree.

> mb i have not considered smth? 

No, it's just that I'm not the most experienced SC writer.

So not an issue.

AnnaShaleva commented on pkg/innerring/internal/metachain/contracts/method_objects.go:
```diff
@@ -0,0 +1,218 @@
+	}
+
+	cnrListRaw := ic.DAO.GetStorageItem(m.ID, append([]byte{containerPlacementPrefix}, cID...))
+	placementI, err := stackitem.Deserialize(cnrListRaw)
```
Ref. https://github.com/nspcc-dev/neo-go/pull/4118.

AnnaShaleva commented on cmd/neofs-node/meta_new.go:
```diff
@@ -0,0 +1,164 @@
+	p2pCfg := c.appCfg.Meta.P2P
+
+	var chainCfg = config.Config{
+		ProtocolConfiguration: config.ProtocolConfiguration{
```
Just for the record: I don't see `Hardforks` being set. It means that all *stable* hardforks are active starting from genesis. It's expected and correct behaviour for meta chain, I'm just letting you know that.

However, NeoGo is continuously adding more hardforks, so if some new hardfork is added, then it may affect states after subsequent NeoGo dependency update in meta chain. So to me, it's better to explicitly set all desired hardfork heights to 0 here.

carpawell commented on pkg/services/sidechain/sidechain.go:
```diff
@@ -0,0 +1,119 @@
+		}
+	}()
+
+	rpcServer := rpcsrv.New(chain, cfg.ApplicationConfiguration.RPC, netServer, nil, log.With(zap.String("subcomponent", "rpc server")), chErr)
```
for Notary -- yes, but what about RPC? or maybe i need to use `(bc *Blockchain) PoolTx`?

carpawell commented on pkg/innerring/internal/metachain/contracts/meta_test.go:
```diff
@@ -0,0 +1,262 @@
+			newContracts = append(newContracts, contract)
+		case *native.Std, *native.Crypto, *native.Oracle:
+		default:
+			panic(fmt.Sprintf("unexpected native contract found: %T", contract))
```
i think it is a feature, if we need more contracts to support at some point in the future (who knows), we'd better know it from tests immediately

AnnaShaleva replied:
Panic is OK to me. Just fetch the latest NeoGo master to properly handle the new native.

carpawell commented on pkg/innerring/internal/metachain/contracts/method_objects.go:
```diff
@@ -0,0 +1,218 @@
+	err = ic.AddNotification(m.Hash, putObjectEvent, stackitem.NewArray([]stackitem.Item{
+		stackitem.NewByteArray(cID),
+		stackitem.NewByteArray(oID),
+		stackitem.NewMapWithValue(metaInfo)}))
```
in the previous implementation -- yes. i have not found anything different yet to implement. SNs are required to get this metainfo to index it (even if they have not received the object itself)

carpawell committed:
go.mod: update neo-go and neofs-contracts

From neo-go side it is useful to have https://github.com/nspcc-dev/neo-go/pull/4117
and https://github.com/nspcc-dev/neo-go/pull/4118. Contracts are updated to keep
deploy code buildable.

Signed-off-by: Pavel Karpy <carpawell@nspcc.ru>

carpawell committed:
wip

Signed-off-by: Pavel Karpy <carpawell@nspcc.ru>

carpawell commented:
PR was split into smaller parts.
